Don't attempt to serialize empty response data.

paulyoung opened the pull request:
This addresses the issue described in #31.

paulyoung committed:
Don't attempt to serialize empty response data.

paulyoung committed:
Add spec and fakes for testing empty response data.

paulyoung commented:
I added a test and some fakes but there's probably a better way to do that.

paulyoung commented on Tests/Fakes/FakeRequest.swift:
```diff
@@ -30,3 +30,19 @@ struct FakeNullDataRequest: Request {
+      return Result.Failure(Result<EmptyResponse, NSError>.error())
+    }
+  }
+}
```
Not sure what I was thinking here, this is the same as `FakeNullDataRequest`.

paulyoung replied:
I guess I was trying to distinguish between null data and data with zero bytes, even though the implementations are currently the same.

paulyoung commented on Tests/Tests/APIClientSpec.swift:
```diff
@@ -66,6 +66,20 @@ class APIClientSpec: QuickSpec {
+                
+                client.performRequest(request) { result = $0 }
+                
+                expect(result!.value!).toEventually(beVoid())
```
Without the `where` clause I added, this will crash. Not ideal.

paulyoung commented:
@tonyd256 @gfontenot any suggestion on how to improve this?

paulyoung commented:
Probably going to take another look at this in the morning.

sidraval commented:
LGTM :+1:

paulyoung commented:
Thanks @sidraval. It definitely still feels half baked on the testing side to me.

sidraval commented:
@paulyoung Perhaps I'm missing something, can you clarify what feels half baked?

paulyoung commented:
I think mainly:
1. There was already a test for null data, but not data with 0 bytes which is the crux of this issue. I added `FakeEmptyDataRequest` to correspond to `FakeNullDataRequest` but they have the same implementation. I feel like maybe they shouldn't and I may have overlooked something there.
2. Similar story for `FakeEmptyResponseRequestPerformer`. It's close to `FakeRequestPerformer` so maybe they could be combined, or it's a sign that some things can be removed from `FakeEmptyResponseRequestPerformer`.
3. The test case I added will crash if there's a regression. I should fix that.

paulyoung commented:
Although for number 3, the other tests do the same...

gfontenot commented:
I'm :+1: on this change in general. I'm also OK with merging these tests as they are, and then trying to refactor these fakes a bit later to clean it up. Fixing this issue was in my mental TODO list (that would _really_ be better not in my head), so I think it's a really good change.

gfontenot commented:
So, basically, @paulyoung if you feel like 80% good with this as it stands, we can merge it and come back to do a refactoring pass later.

paulyoung commented:
Sounds good!

paulyoung commented:
@gfontenot good to go? :bow: :pray:

gfontenot commented:
@paulyoung yup yup! Merged in as 5911762. Thanks!
